src/utils/FileDescriptor.cpp

```diff
@@ -0,0 +1,77 @@
+#include <hyprutils/utils/FileDescriptor.hpp>
+#include <fcntl.h>
+#include <sys/poll.h>
+#include <unistd.h>
+#include <utility>
+
+using namespace Hyprutils::Utils;
+
+CFileDescriptor::CFileDescriptor(int fd) : m_fd(fd) {}
+
+CFileDescriptor::CFileDescriptor(CFileDescriptor&& other) : m_fd(std::exchange(other.m_fd, -1)) {}
+
+CFileDescriptor& CFileDescriptor::operator=(CFileDescriptor&& other) {
+    if (this == &other)
+        return *this;
+
+    reset();
+    m_fd = std::exchange(other.m_fd, -1);
+    return *this;
+}
+
+CFileDescriptor::~CFileDescriptor() {
+    reset();
+}
+
+bool CFileDescriptor::isValid() const {
+    return m_fd != -1;
+}
+
+int CFileDescriptor::get() const {
+    return m_fd;
+}
+
+int CFileDescriptor::take() {
+    return std::exchange(m_fd, -1);
+}
+
+void CFileDescriptor::reset() {
+    if (m_fd != -1) {
+        close(m_fd);
+        m_fd = -1;
+    }
+}
+
+CFileDescriptor CFileDescriptor::duplicate() const {
+    if (m_fd == -1)
+        return {};
+
+    return CFileDescriptor{fcntl(m_fd, F_DUPFD_CLOEXEC, 0)};
+}
+
+bool CFileDescriptor::isClosed() const {
+    return isClosed(m_fd);
+}
+
+bool CFileDescriptor::isReadable() const {
+    return isReadable(m_fd);
+}
+
+bool CFileDescriptor::isClosed(int fd) {
+    pollfd pfd = {
+        .fd      = fd,
+        .events  = POLLIN,
+        .revents = 0,
+    };
+
+    if (poll(&pfd, 1, 0) < 0)
+        return true;
+
+    return pfd.revents & (POLLHUP | POLLERR);
+}
+
+bool CFileDescriptor::isReadable(int fd) {
+    pollfd pfd = {.fd = fd, .events = POLLIN, .revents = 0};
```

Comment:
what for

Reply:
in kwin isReadable/isClosed is used in a few protocols where they dont have the ownership but the fd might be closed/not readable on certain events/occations. like in securitycontext according to the protocol the close_fd is a FD closed by the client when the compositor should stop accepting new connections on listen_fd. so added it ahead of time

Reply:
mhm, ok